# Homework 3  - Tryout pandas and numpy

Copy this notebook. Rename it as: YOURNAME-Homework-panda-numpy-Sept-2017

with your name replacing yourname and  replaced with the date you submit or to the last part.

Do the homeworks problems below.

Upload your completed jupyter notebook to your github site and send me the url via the elearning site as your homework submission. Please note that you need to do your own 3.4, and 3.5 if you share data and analysis from others then you cannot get more than a 3.  

## Homework 3.1

### 3.1.a
Load the data from: http://opendata.dc.gov/datasets that I have include in this github   
into a dataframe.  ( The file has been is available in directory ./data/ccp_current_csv.csv )

In [219]:
import pandas as pd

data = pd.read_csv("data/ccp_current_csv.csv")
data.head()


STREET QUADRANT  FROMINTERSECTION              TOINTERSECTION  \
0           13th St       NW        Kennedy St               Longfellow St   
1              S St       NW  North Capitol St                      3rd St   
2  Rhode Island Ave       NE  Brentwood Rd (W)            Brentwood Rd (M)   
3           Mill Rd       NW           27th St                  Cul De Sac   
4  Rhode Island Ave       NE  Brentwood Rd (M)  14th St  Evarts St  Montan   

   NOOFBLOCKS                 WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  \
0           1  Pavement Resurface/restoration          2004               100   
1           3                   Regular Cover          1999               100   
2           1  Pavement Resurface/restoration          2004               100   
3           1  Pavement Resurface/restoration          2004               100   
4           2  Pavement Resurface/restoration          2005               100   

      STATUS  MILES      ...             YCOORD   LATITUDE  LONGITUDE  \
0  Completed   0.07      ...       143205.11981  38.956747 -77.029694   
1  Completed   0.33      ...       138468.90256  38.914085 -77.012411   
2  Completed   0.09      ...       139574.10659  38.924041 -76.987221   
3  Completed   0.07      ...       138132.90198  38.911046 -77.054077   
4  Completed   0.16      ...       139657.25318  38.924790 -76.985068   

           PROJECTNOTE                                        PROJECTNAME  \
0                  NaN                                             FedAid   
1                  NaN                                                NaN   
2                  NaN           Fedral Aid Citywide Pavement Restoration   
3  POKA-2003-B-0048-JJ  Citywide Pavement Restoration Contract (Local ...   
4                  NaN                                                NaN   

          PROJECTMGR  PROJECTID        TEAMREP                 EMAIL  \
0    Paul Stephenson       5834        O. Hill                   NaN   
1                NaN       2467            NaN                   NaN   
2  Mohamed Abdullahi       5486   Said Cherifi   said.cherifi@dc.gov   
3    Maduabuchi Udeh       5796  Michael Jelen  Michael.Jelen@dc.gov   
4                NaN       8237            NaN                   NaN   

   CONTACTPHONE  
0  202-671-4591  
1           NaN  
2  202-671-4611  
3  202-671-4542  
4           NaN  

[5 rows x 27 columns]

### 3.1.a  what is its shape and what does that mean?

In [220]:
data.shape

(465, 27)

Here 465 is number of different data entries(rows) and 27 is the total number of features in each data entry(columns)which includes class(predict) attribute also.

## Homework 3.2
 What are the number of rows in each 'QUADRANT' ?

In [222]:
# using value_counts() function we can get count of each QUADRANT
data['QUADRANT'].value_counts()

NW    195
NE    163
SE     84
BN     12
SW     11
Name: QUADRANT, dtype: int64

In [223]:
import numpy as np

a=np.array([1,2,3,4],dtype=float)
b=np.array([5,6,7,8],dtype=float)

## Homework 3.3 - Array math demonstration 
For two arrarys  

a= [1,2,3,4]  type=float

b= [5,6,7,8]  type=float

Peform the following array operations using numpy   
( show both operational use of numpy and functional (example addition operation => + vs addition function => numbpy.add() )

### 3.3.1 addition a+b 

In [224]:
# using Operator
print("Operator:",a+b)

# using numpy add function
print("Numpy Add:",np.add(a,b))


Operator: [  6.   8.  10.  12.]
Numpy Add: [  6.   8.  10.  12.]


### 3.3.2 subtraction a-b

In [225]:
# using Operator
print("Operator:",a-b)

# using numpy subtract function
print("Numpy subtract:",np.subtract(a,b))

Operator: [-4. -4. -4. -4.]
Numpy subtract: [-4. -4. -4. -4.]


### 3.3.3 multiplication a*b

In [226]:
# using Operator
print("Operator:",a*b)

# using numpy multiply function
print("Numpy subtract:",np.multiply(a,b))

Operator: [  5.  12.  21.  32.]
Numpy subtract: [  5.  12.  21.  32.]


### 3.3.4 divsion a/b

In [227]:
# using Operator
print("Operator:",a/b)

# using numpy Divide function
print("Numpy Divide:",np.divide(a,b))

Operator: [ 0.2         0.33333333  0.42857143  0.5       ]
Numpy Divide: [ 0.2         0.33333333  0.42857143  0.5       ]


### 3.3.5 modulo  a%b

In [228]:
# using Operator
print("Operator:",a%b)

# using numpy modulo function
print("Numpy Divide:",np.mod(a,b))

Operator: [ 1.  2.  3.  4.]
Numpy Divide: [ 1.  2.  3.  4.]


### 3.3.6 power  a^b

In [229]:
# using Operator
print("Operator:",a**b)

# using numpy power function
print("Numpy Divide:",np.power(a,b))

Operator: [  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]
Numpy Divide: [  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]


## Homework 3.4
Find your own data and load it into a dataframe

Dataset : Heart Disease

link:     http://archive.ics.uci.edu/ml/datasets/heart+Disease

Description:

The following dataset has been used to determine heart disease in the patient. While '0' indicates no heart disease, '1','2','3','4' indicates some kind of heart disease exist in patients.

The dataset has 14 attributes including class attribute('NUM').

The dataset has some missing values indicated by '?'

In [230]:
# load the data using pandas read_csv function and replace the missing value indicated by '?' into NaN
hd_data=pd.read_csv("data/Heart_Disease.csv",na_values='?')
hd_data.head()

AGE  SEX  CP  TRESTBPC  CHOL  FBS  RESTECG  THALACH  EXANG  OLDPEAK  SLOPE  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    CA  THAL  NUM  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0

## Homework 3.5
Provide an interesting analysis of the data columns ( frequency or averages )

In [231]:
# get the number of rows and number of coulmns(features including class attribute) in dataset
hd_data.shape

(303, 14)

In [232]:
# use isnull function to determine if the dataset has any missing value
# any() function outputs 'True' if any value in the particular coulmn has null value 
# and 'False' if and only if entire column has no Null value 
hd_data.isnull().any()

AGE         False
SEX         False
CP          False
TRESTBPC    False
CHOL        False
FBS         False
RESTECG     False
THALACH     False
EXANG       False
OLDPEAK     False
SLOPE       False
CA           True
THAL         True
NUM         False
dtype: bool

In [233]:
# Sum() function along with isnull() function can be used to get count of null values in each column
# second sum() function is used to add null values of all the coulmns
hd_data.isnull().sum().sum()

6

In [234]:
# display rows with Null values
hd_data[hd_data.CA.isnull() | hd_data.THAL.isnull()]

AGE  SEX  CP  TRESTBPC  CHOL  FBS  RESTECG  THALACH  EXANG  OLDPEAK  \
87    53    0   3       128   216    0        2      115      0      0.0   
166   52    1   3       138   223    0        0      169      0      0.0   
192   43    1   4       132   247    1        2      143      1      0.1   
266   52    1   4       128   204    1        0      156      1      1.0   
287   58    1   2       125   220    0        0      144      0      0.4   
302   38    1   3       138   175    0        0      173      0      0.0   

     SLOPE   CA  THAL  NUM  
87       1  0.0   NaN    0  
166      1  NaN   3.0    0  
192      2  NaN   7.0    1  
266      2  0.0   NaN    2  
287      2  NaN   7.0    0  
302      1  NaN   3.0    0

In [235]:
# missing values can be replaced by specific value or mean of that particular column etc..
hd_data['CA'].fillna(0.0, inplace=True)
hd_data['THAL'].fillna(hd_data['THAL'].mean(), inplace=True)

In [236]:
# check again if there are any Null values
hd_data.isnull().any()

AGE         False
SEX         False
CP          False
TRESTBPC    False
CHOL        False
FBS         False
RESTECG     False
THALACH     False
EXANG       False
OLDPEAK     False
SLOPE       False
CA          False
THAL        False
NUM         False
dtype: bool

In [237]:
# for the purpose of analysis, replacing the values 2, 3, 4 into 1 to classify 
# all of them as presence of heart disease
hd_data['NUM'].replace([2,3,4], 1, inplace=True)

#### Analysis

In [238]:
# calculate the number of patient suffering heart disease and not suffering heart disease
hd_data['NUM'].value_counts()

0    164
1    139
Name: NUM, dtype: int64

Out of 303, 164 people suffer some kind of heart problem whereas 139 people dont have any heart problem.

In [239]:
# get the number of males and females in the dataset
hd_data['SEX'].value_counts()

1    206
0     97
Name: SEX, dtype: int64

Out of 303, 207 people are male 97 are female.This indicates there are more males than females in this dataset.

In [240]:
# calculate the number of male and female suffering from heart disease
# we can use groupby function to know, how many male and female patient suffer from heart disease
result=hd_data.groupby(['SEX','NUM'])['SEX'].count()
print(result)
print(result/len(hd_data)*100)

SEX  NUM
0    0       72
     1       25
1    0       92
     1      114
Name: SEX, dtype: int64
SEX  NUM
0    0      23.762376
     1       8.250825
1    0      30.363036
     1      37.623762
Name: SEX, dtype: float64


From the first and second result, we can see, out of 97 females, 25 females(8.25%) suffer from Heart disease and out of 206 males, 114 males(37%) suffer from heart disease.
From the above results, males are more likely to suffer from heart disease than females.


In [241]:
# let us see, if age determines the heart disease or not
# first let us know the mean age of the dataset where there is some kind of heart disease exist
# we can use pandas mean function to calculate mean
hd_data['AGE'][hd_data['NUM']==1].mean()


56.62589928057554

From the above results,aged people are more likely to suffer from some kind of heart disease.

In [242]:
# more analysis using age attribute
# we can put ages into bins using pandas cut function  
# and then use value_counts() function to get count of each bin
bins=[0,30,50,80]
hd_data['AGE_BIN'] = pd.cut(hd_data['AGE'], bins=bins)
hd_data['AGE_BIN'].value_counts()

(50, 80]    209
(30, 50]     93
(0, 30]       1
Name: AGE_BIN, dtype: int64

In [243]:
# using groupby function , we can group AGE_BIN based on heart disease or no heart disease category
hd_data.groupby(['AGE_BIN','NUM'])['AGE_BIN'].count()

AGE_BIN   NUM
(0, 30]   0        1
(30, 50]  0       64
          1       29
(50, 80]  0       99
          1      110
Name: AGE_BIN, dtype: int64

From the above results, its more likely for people after age 50 suffer from some kind of heart disease.

In [244]:
# let us see if cholestrol level is directly responsible for heart disease
# get the number of patient with high chol and suffering heart disease 
sum(hd_data.loc[hd_data['CHOL'] > 250]['NUM']==1)

69

Above 69 indicates, only 69 people out of 303 are having some kind of heart diease.Hence,knowing just cholestrol level would not help in determining heart disease.


In [245]:
# we can also use corr() function to determine if there is any correlatin between data
hd_data[['NUM','CHOL','AGE','THAL']].corr()

NUM      CHOL       AGE      THAL
NUM   1.000000  0.085164  0.223120  0.523928
CHOL  0.085164  1.000000  0.208950  0.014190
AGE   0.223120  0.208950  1.000000  0.127368
THAL  0.523928  0.014190  0.127368  1.000000

Above, we can see CHOL and NUM are not highly correlated to each other, AGE is better than CHOL and, THAL and NUM are highly correlated.

In [246]:
hd_data.groupby(['THAL','NUM'])['THAL'].count()

THAL      NUM
3.000000  0      129
          1       37
4.734219  0        1
          1        1
6.000000  0        6
          1       12
7.000000  0       28
          1       89
Name: THAL, dtype: int64

As we can see , people with THAL value 6 and 7 are suffering from some kind of heart disease.

In [247]:
# we can use describe() function to get overall picture about dataset 
hd_data.describe()

AGE         SEX          CP    TRESTBPC        CHOL         FBS  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          RESTECG     THALACH       EXANG     OLDPEAK       SLOPE          CA  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.663366   
std      0.994971   22.875003    0.469794    1.161075    0.616226    0.934375   
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000   
75%      2.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    3.000000    3.000000   

             THAL         NUM  
count  303.000000  303.000000  
mean     4.734219    0.458746  
std      1.933272    0.499120  
min      3.000000    0.000000  
25%      3.000000    0.000000  
50%      3.000000    0.000000  
75%      7.000000    1.000000  
max      7.000000    1.000000

In [248]:
# we can even use agg function instead of describe() to each column (feature)  and specify set of operation
# need to be applied to each column
hd_data.agg(['mean', 'min','max'])

AGE       SEX        CP    TRESTBPC        CHOL       FBS  \
max   77.000000  1.000000  4.000000  200.000000  564.000000  1.000000   
mean  54.438944  0.679868  3.158416  131.689769  246.693069  0.148515   
min   29.000000  0.000000  1.000000   94.000000  126.000000  0.000000   

       RESTECG     THALACH     EXANG   OLDPEAK    SLOPE        CA      THAL  \
max   2.000000  202.000000  1.000000  6.200000  3.00000  3.000000  7.000000   
mean  0.990099  149.607261  0.326733  1.039604  1.60066  0.663366  4.734219   
min   0.000000   71.000000  0.000000  0.000000  1.00000  0.000000  3.000000   

           NUM   AGE_BIN  
max   1.000000  (50, 80]  
mean  0.458746       NaN  
min   0.000000   (0, 30]

In [249]:
# some Machine learning algorithm might not work correctly if features are not normalized 
# we can easily normalize the data using pandas functions
import numpy as np
del hd_data['AGE_BIN']
#hd_data_nor = hd_data.apply(lambda x: (x - np.mean(x)) / (np.max(x) - np.min(x)))
hd_data_nor = (hd_data - hd_data.min()) / (hd_data.max() - hd_data.min())
hd_data_nor.agg(['mean', 'min','max']) 

AGE       SEX        CP  TRESTBPC      CHOL       FBS  RESTECG  \
mean  0.529978  0.679868  0.719472  0.355564  0.275555  0.148515  0.49505   
min   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.00000   
max   1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.00000   

       THALACH     EXANG   OLDPEAK    SLOPE        CA      THAL       NUM  
mean  0.600055  0.326733  0.167678  0.30033  0.221122  0.433555  0.458746  
min   0.000000  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000  
max   1.000000  1.000000  1.000000  1.00000  1.000000  1.000000  1.000000

As we can see now, all the features are values are between 0 and 1